src/rate.rs

```diff
@@ -0,0 +1,213 @@
+use anyhow::Context;
+use num::{BigUint, Integer, ToPrimitive};
+use std::convert::TryFrom;
+use std::iter::FromIterator;
+use std::str::FromStr;
+
+/// Represent a rate. Note this is designed to support Bitcoin/Dai buy and sell rates (Bitcoin being in the range of 10k-100kDai)
+/// A rate has a maximum precision of 9 digits after the decimal
+// rate = self.0 * 10e-9
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Rate(u64);
+
+impl Rate {
+    pub const PRECISION: u16 = 9;
+
+    /// integer = rate * 10ePRECISION
+    pub fn new(integer: u64) -> anyhow::Result<Self> {
+        Ok(Rate(integer))
+    }
+
+    /// integer = rate * 10ePRECISION
+    pub fn integer(&self) -> BigUint {
+        BigUint::from(self.0)
+    }
+}
+
+impl TryFrom<f64> for Rate {
+    type Error = anyhow::Error;
+
+    fn try_from(rate: f64) -> Result<Self, Self::Error> {
+        if rate.is_sign_negative() {
+            anyhow::bail!("Rate must be positive");
+        }
+
+        if !rate.is_finite() {
+            anyhow::bail!("Rate must be finite")
+        }
+
+        let mut rate = rate.to_string();
+        let decimal_index = rate.find('.');
+        let mantissa = match decimal_index {
+            None => String::new(),
+            Some(decimal_index) => {
+                let mantissa = rate.split_off(decimal_index + 1);
+                if mantissa.len() > Self::PRECISION as usize {
+                    anyhow::bail!(format!(
+                        "Precision of the rate is too high (max is {}).",
+                        Self::PRECISION
+                    ))
+                }
+                rate.truncate(rate.len() - 1); // Removes the trailing decimal point
+                mantissa
+            }
+        };
+
+        let mantissa_length = mantissa.len();
+        let integer = rate;
+
+        let zeros = vec!['0'].repeat(Self::PRECISION as usize - mantissa_length);
+        let zeros = String::from_iter(zeros.into_iter());
+        let integer = u64::from_str(&format!("{}{}{}", integer, mantissa, zeros))
+            .context("Rate is unexpectedly large")?;
+        Rate::new(integer)
+    }
+}
+
+/// Spread: percentage to be added on top of a rate or amount with
+/// a maximum precision of 2 decimals
+#[derive(Clone, Copy, Debug)]
+pub struct Spread(u16);
+
+impl Spread {
+    /// Input is the spread in permyriad (per ten thousand):
+    /// 5% => 500 permyriad
+    /// 23.14% => 2314 permyriad
+    /// 0.001% => Not allowed
+    /// 200% => Not allowed
+    pub fn new(permyriad: u16) -> anyhow::Result<Spread> {
+        if permyriad > 10000 {
+            anyhow::bail!("Spread must be between 0% and 100%");
+        }
+
+        Ok(Spread(permyriad))
+    }
+
+    pub fn apply(&self, rate: Rate) -> anyhow::Result<Rate> {
```

Comment:
:upside_down_face: I think it would read better if this function were on `Rate` and took a `Spread` as an argument:

```rust
rate.apply(spread)
```

Reply:
I am going to stop here :)